### PR TITLE
refactor: add context api

### PR DESCRIPTION
**Pull request by cyberhck**
## Pull Request Submission Checklist

Please confirm that you have done the following before requesting reviews:

- [x] I have confirmed that the PR type is appropriate for the change I am making according to
  the [Honest Pull Request and Commit Message Naming Conventions](https://www.notion.so/honestbank/Pull-Request-and-Commit-Message-Naming-Conventions-bd97f2cbb34c4c73b1ff3a3e384b850c).
- [x] I have typed an adequate description that explains **why** I am making this change.
- [x] I have installed and run standard pre-commit hooks that lints and validates my code.

### Description

* This adds context API so that we can pass values around.
* This also now makes it such that only the core functionality along with retries, dead lettering, and middlewares are first-class citizens of KP.
* things like backoff, tracing, measurements, logs, etc are second-class citizens and will be done through middleware and the KP won't expose API directly.
* context will also allow us to get things like retry count, tracing etc in our worker.

- - -
This change is [Reviewable](https://reviewable.io/reviews/honestbank/kp/20)

### CODE REVIEW

**Review by kelu-diao — APPROVED** [review]

File: v2/internal/middleware/middleware.go [hunk 1]
```diff
@@ -31,15 +33,15 @@ func (r *stack[IN, OUT]) AddMiddleware(mw Middleware[IN, OUT]) {
 	r.middlewares = append(r.middlewares, mw)
 }
 
-func (r *stack[IN, OUT]) Process(options IN) OUT {
-	var next func(item IN) OUT = nil
+func (r *stack[IN, OUT]) Process(ctx context.Context, options IN) OUT {
+	var next func(c context.Context, item IN) OUT = nil
 	middlewares := append(r.middlewares)
```
Comment (kelu-diao) [inline v2/internal/middleware/middleware.go:38]:
```suggestion
	middlewares := r.middlewares
```

File: v2/internal/middleware/middleware.go [hunk 2]
```diff
@@ -31,15 +33,15 @@ func (r *stack[IN, OUT]) AddMiddleware(mw Middleware[IN, OUT]) {
...
-	next = func(item IN) OUT {
+	next = func(c context.Context, item IN) OUT {
 		nextMw := middlewares[0]
 		middlewares = middlewares[1:]
-		return nextMw.Process(item, next)
+		return nextMw.Process(c, item, next)
 	}
-	return next(options)
+	return next(ctx, options)
```
Comment (kelu-diao) [inline v2/internal/middleware/middleware.go:44]: minor: it's a bit hard to read when everything is named as next, it's probably better to rename `nextMw` to `currentMw`

**Review by kelu-diao — APPROVED** [review]
Great work! just a few minor comments on readability.

File: v2/internal/middleware/middleware.go (hunk 1 above)
Comment (cyberhck) [reply v2/internal/middleware/middleware.go:38]: This was actually a bug in the code, I meant to make a duplicate, not a reference copy, I'm fixing using copy method.

File: v2/internal/middleware/middleware.go (hunk 2 above)
Comment (cyberhck) [reply v2/internal/middleware/middleware.go]: renamed :) thanks.

**Review by bludot — APPROVED** [review]
LGTM but get a 3rd review pls

File: v2/internal/middleware/middleware.go (hunk 1 above)
Comment (kelu-diao) [reply v2/internal/middleware/middleware.go]: This was fine. Assigning `r.middlewares` to `middlewares` makes a shallow copy of the slice, hence the index changes like `middlewares = middlewares[1:]` doesn't affect the `r.middlewares`.

**Pull request merged**